modules/lang-mustache/src/main/java/org/elasticsearch/action/search/template/TransportSearchTemplateAction.java

```diff
@@ -0,0 +1,122 @@
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.index.query.QueryParseContext;
+import org.elasticsearch.indices.query.IndicesQueriesRegistry;
+import org.elasticsearch.script.ExecutableScript;
+import org.elasticsearch.script.Script;
+import org.elasticsearch.script.ScriptService;
+import org.elasticsearch.script.mustache.MustacheScriptEngineService;
+import org.elasticsearch.search.aggregations.AggregatorParsers;
+import org.elasticsearch.search.builder.SearchSourceBuilder;
+import org.elasticsearch.search.suggest.Suggesters;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+
+import java.io.IOException;
+
+import static java.util.Collections.emptyMap;
+import static org.elasticsearch.script.ScriptContext.Standard.SEARCH;
+
+public class TransportSearchTemplateAction extends HandledTransportAction<SearchTemplateRequest, SearchTemplateResponse> {
+
+    private static final String TEMPLATE_LANG = MustacheScriptEngineService.NAME;
+
+    private final ClusterService clusterService;
+    private final ScriptService scriptService;
+    private final TransportSearchAction searchAction;
+    private final IndicesQueriesRegistry queryRegistry;
+    private final AggregatorParsers aggsParsers;
+    private final Suggesters suggesters;
+
+    @Inject
+    public TransportSearchTemplateAction(Settings settings, ThreadPool threadPool, TransportService transportService,
+                                         ActionFilters actionFilters, IndexNameExpressionResolver resolver,
+                                         ClusterService clusterService, ScriptService scriptService,
+                                         TransportSearchAction searchAction, IndicesQueriesRegistry indicesQueryRegistry,
+                                         AggregatorParsers aggregatorParsers, Suggesters suggesters) {
+        super(settings, SearchTemplateAction.NAME, threadPool, transportService, actionFilters, resolver, SearchTemplateRequest::new);
+        this.clusterService = clusterService;
+        this.scriptService = scriptService;
+        this.searchAction = searchAction;
+        this.queryRegistry = indicesQueryRegistry;
+        this.aggsParsers = aggregatorParsers;
+        this.suggesters = suggesters;
+    }
+
+    @Override
+    protected void doExecute(SearchTemplateRequest request, ActionListener<SearchTemplateResponse> listener) {
+        final SearchTemplateResponse response = new SearchTemplateResponse();
+        try {
+            Script script = new Script(request.getScript(), request.getScriptType(), TEMPLATE_LANG, request.getScriptParams());
+            ExecutableScript executable = scriptService.executable(script, SEARCH, emptyMap(), clusterService.state());
+
+            BytesReference source = (BytesReference) executable.run();
+            response.setSource(source);
+
+            if (request.isSimulate() == false) {
```

Comment:
maybe instead of the large if / else clauses, we could write:
```java
if (request.isSimulate()) {
   listener.onResponse(response);
   return;
}
// continue with delegating to search api
```

I think this is cleaner?